File: frontend/android/app/src/main/java/app/sasl/plugin/WaveMeshPlugin.java
package app.sasl.plugin;

import android.content.Context;
import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import app.sasl.wavemesh.WaveMeshNativeService;

@CapacitorPlugin(name = "WaveMeshPlugin")
public class WaveMeshPlugin extends Plugin {
    private static final String TAG = "WaveMeshPlugin";
    private WaveMeshNativeService waveMeshService;
    
    @Override
    public void load() {
        Context context = getBridge().getContext();
        waveMeshService = WaveMeshNativeService.getInstance(context);
        waveMeshService.setCallback(new WaveMeshNativeService.WaveMeshCallback() {
            @Override
            public void onPeerDiscovered(String deviceId, String name, String connectionType, int distance) {
                JSObject peer = new JSObject();
                peer.put("deviceId", deviceId); peer.put("name", name);
                peer.put("connectionType", connectionType); peer.put("distance", distance);
                notifyListeners("peerDiscovered", peer);
            }
            @Override
            public void onPeerConnected(String deviceId, String name) {
                JSObject peer = new JSObject();
                peer.put("deviceId", deviceId); peer.put("name", name);
                notifyListeners("peerConnected", peer);
            }
            @Override
            public void onMessageReceived(String from, String text) {
                JSObject msg = new JSObject();
                msg.put("from", from); msg.put("text", text);
                notifyListeners("messageReceived", msg);
            }
            @Override
            public void onICECandidate(String from, String candidate) {}
            @Override
            public void onStatusChanged(String status) {
                JSObject s = new JSObject(); s.put("status", status);
                notifyListeners("statusChanged", s);
            }
        });
        Log.d(TAG, "WaveMeshPlugin loaded");
    }
    
    @PluginMethod public void setIdentity(PluginCall call) {
        waveMeshService.setIdentity(call.getString("id",""), call.getString("username","User"));
        call.resolve();
    }
    @PluginMethod public void startAdvertising(PluginCall call) {
        waveMeshService.startAdvertising(call.getString("username", "SaslUser"));
        call.resolve();
    }
    @PluginMethod public void stopAdvertising(PluginCall call) {
        waveMeshService.stopAdvertising(); call.resolve();
    }
    @PluginMethod public void startBLEScan(PluginCall call) {
        waveMeshService.startBLEScan(); call.resolve();
    }
    @PluginMethod public void stopBLEScan(PluginCall call) {
        waveMeshService.stopBLEScan(); call.resolve();
    }
    @PluginMethod public void connectToPeer(PluginCall call) {
        waveMeshService.connectToPeer(call.getString("deviceAddress",""));
        call.resolve();
    }
    @PluginMethod public void sendMessage(PluginCall call) {
        waveMeshService.sendMessage(call.getString("deviceAddress",""), call.getString("message",""));
        call.resolve();
    }
    @PluginMethod public void getCapabilities(PluginCall call) {
        JSObject caps = new JSObject();
        caps.put("bleReady", waveMeshService.isBleReady());
        caps.put("isAdvertising", waveMeshService.isAdvertising());
        caps.put("isScanning", waveMeshService.isScanning());
        call.resolve(caps);
    }
    @PluginMethod public void stop(PluginCall call) {
        waveMeshService.stop(); call.resolve();
    }


        @PluginMethod public void startRelayService(PluginCall call) {
        Intent intent = new Intent(getContext(), EchoRelayService.class);
        intent.putExtra("username", call.getString("username", "RelayNode"));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            getContext().startForegroundService(intent);
        } else {
            getContext().startService(intent);
        }
        call.resolve();
    }

    @PluginMethod public void stopRelayService(PluginCall call) {
        Intent intent = new Intent(getContext(), EchoRelayService.class);
        getContext().stopService(intent);
        call.resolve();
    }
}

File: frontend/ios/App/App/WaveMeshPlugin.swift
import Foundation
import Capacitor

@objc(WaveMeshPlugin)
public class WaveMeshPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WaveMeshPlugin"
    public let jsName = "WaveMeshPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setIdentity", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startBLEScan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopBLEScan", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "connectToPeer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendOverBLE", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMultipeerDiscovery", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendOverMultipeer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCapabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise),
    ]
    
    private let waveMesh = WaveMeshNativeService.shared
    
    override public func load() {
        waveMesh.onPeerDiscovered = { [weak self] deviceId, name, connectionType, distance in
            self?.notifyListeners("peerDiscovered", data: [
                "deviceId": deviceId,
                "name": name,
                "connectionType": connectionType,
                "distance": distance
            ])
        }
        
        waveMesh.onPeerConnected = { [weak self] deviceId, name in
            self?.notifyListeners("peerConnected", data: [
                "deviceId": deviceId,
                "name": name
            ])
        }
        
        waveMesh.onMessageReceived = { [weak self] from, text in
            self?.notifyListeners("messageReceived", data: [
                "from": from,
                "text": text
            ])
        }
        
        waveMesh.onICECandidate = { [weak self] from, candidate in
            self?.notifyListeners("iceCandidate", data: [
                "from": from,
                "candidate": candidate
            ])
        }
        
        print("📱 WaveMeshPlugin iOS loaded")
    }
    
    @objc func setIdentity(_ call: CAPPluginCall) {
        let id = call.getString("id") ?? "unknown"
        let username = call.getString("username") ?? "User"
        waveMesh.setIdentity(id: id, username: username)
        call.resolve()
    }
    
    @objc func startBLEScan(_ call: CAPPluginCall) {
        waveMesh.startBLEScan()
        waveMesh.startMultipeerDiscovery()
        call.resolve()
    }
    
    @objc func stopBLEScan(_ call: CAPPluginCall) {
        waveMesh.stopBLEScan()
        waveMesh.stopMultipeer()
        call.resolve()
    }
    
    @objc func connectToPeer(_ call: CAPPluginCall) {
        if let deviceAddress = call.getString("deviceAddress") {
            waveMesh.connectToPeer(deviceAddress)
        }
        call.resolve()
    }
    
    @objc func sendOverBLE(_ call: CAPPluginCall) {
        let peerAddress = call.getString("peerAddress") ?? ""
        let type = call.getString("type", "MSG")
        let data = call.getString("data", "")
        waveMesh.sendOverBLE(peerAddress: peerAddress, type: type, data: data)
        call.resolve()
    }
    
    @objc func startMultipeerDiscovery(_ call: CAPPluginCall) {
        waveMesh.startMultipeerDiscovery()
        call.resolve()
    }
    
    @objc func sendOverMultipeer(_ call: CAPPluginCall) {
        let text = call.getString("text", "")
        waveMesh.sendOverMultipeer(text)
        call.resolve()
    }
    
    @objc func getCapabilities(_ call: CAPPluginCall) {
        call.resolve(waveMesh.getCapabilities())
    }
    
    @objc func stop(_ call: CAPPluginCall) {
        waveMesh.stopBLEScan()
        waveMesh.stopMultipeer()
        call.resolve()
    }
}
    @objc func startAdvertising(_ call: CAPPluginCall) {
        let username = call.getString("username") ?? "SaslUser"
        waveMesh.startAdvertising(username: username)
        call.resolve()
    }
    
    @objc func stopAdvertising(_ call: CAPPluginCall) {
        waveMesh.stopAdvertising()
        call.resolve()
    }
